crates/reward/src/lib.rs

```diff
@@ -0,0 +1,269 @@
+    pub type TotalRewards<T: Config<I>, I: 'static = ()> = StorageValue<_, T::Balance, ValueQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn reward_per_token)]
+    pub type RewardPerToken<T: Config<I>, I: 'static = ()> = StorageValue<_, SignedFixedPoint<T, I>, ValueQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn stake)]
+    pub type Stake<T: Config<I>, I: 'static = ()> =
+        StorageMap<_, Blake2_128Concat, T::AccountId, T::Balance, ValueQuery>;
+
+    #[pallet::storage]
+    pub type RewardTally<T: Config<I>, I: 'static = ()> =
+        StorageMap<_, Blake2_128Concat, T::AccountId, SignedFixedPoint<T, I>, ValueQuery>;
+
+    #[pallet::pallet]
+    pub struct Pallet<T, I = ()>(_);
+
+    // The pallet's dispatchable functions.
+    #[pallet::call]
+    impl<T: Config<I>, I: 'static> Pallet<T, I> {}
+}
+
+// "Internal" functions, callable by code.
+impl<T: Config<I>, I: 'static> Pallet<T, I> {
+    fn inner_to_balance(x: Inner<T, I>) -> Result<T::Balance, DispatchError> {
+        let y = UniqueSaturatedInto::<u128>::unique_saturated_into(x);
+        TryInto::<T::Balance>::try_into(y).map_err(|_| Error::<T, I>::TryIntoIntError.into())
+    }
+
+    fn balance_to_inner(x: T::Balance) -> Result<Inner<T, I>, DispatchError> {
+        let y = TryInto::<u128>::try_into(x).map_err(|_| Error::<T, I>::TryIntoIntError)?;
+        TryInto::<Inner<T, I>>::try_into(y).map_err(|_| Error::<T, I>::TryIntoIntError.into())
+    }
+
+    fn balance_to_fixed(x: T::Balance) -> Result<SignedFixedPoint<T, I>, DispatchError> {
+        let unsigned_fixed_point = SignedFixedPoint::<T, I>::checked_from_integer(Self::balance_to_inner(x)?)
+            .ok_or(Error::<T, I>::TryIntoIntError)?;
+        Ok(unsigned_fixed_point)
+    }
+
+    pub fn participants() -> Vec<(T::AccountId, T::Balance)> {
+        <Stake<T, I>>::iter().collect()
+    }
+}
+
+pub trait Rewards<AccountId> {
+    /// The balance type for a currency.
+    type Balance: AtLeast32BitUnsigned + FullCodec + Copy + MaybeSerializeDeserialize + Debug + Default;
+
+    fn get_stake(account_id: &AccountId) -> Self::Balance;
+
+    fn deposit_stake(account_id: &AccountId, amount: Self::Balance) -> Result<(), DispatchError>;
+
+    fn distribute(reward: Self::Balance) -> Result<Self::Balance, DispatchError>;
+
+    fn compute_reward(account_id: &AccountId) -> Result<Self::Balance, DispatchError>;
+
+    fn withdraw_stake(account_id: &AccountId, amount: Self::Balance) -> Result<(), DispatchError>;
+
+    fn withdraw_reward(account_id: &AccountId) -> Result<Self::Balance, DispatchError>;
+}
+
+impl<T: Config<I>, I: 'static> Rewards<T::AccountId> for Pallet<T, I>
+where
+    T::Balance: MaybeSerializeDeserialize + Debug,
+{
+    type Balance = T::Balance;
+
+    fn get_stake(account_id: &T::AccountId) -> Self::Balance {
+        Self::stake(account_id)
+    }
+
+    fn deposit_stake(account_id: &T::AccountId, amount: Self::Balance) -> Result<(), DispatchError> {
```

Comment:
Not sure exactly, but if this function is tracking the total collateral locked and the individual collateral locked at a certain point in time, wouldn't it be better to query the vault registry and the staked relayer registry for the current values instead of tracking the same state across two different pallets?

Reply:
Ah, right. Makes sense then!